program/src/processor.rs

```diff
@@ -5313,6 +5318,286 @@ impl Processor {
+                &open_orders_ais,
+                bank_set.token_index,
+            )?;
+            // TODO log
+        }
+        let post_health = health_cache.get_health(&mango_group, HealthType::Init);
+        // If an account is in reduce_only mode, health must only go up
+        check!(
+            post_health >= ZERO_I80F48 || (reduce_only && post_health >= pre_health),
+            MangoErrorCode::InsufficientFunds
+        )?;
+
+        Ok(())
+    }
+
+    fn flash_loan(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        liquidity_amount: u64,
+        cpi_data: Vec<u8>,
+    ) -> MangoResult {
+        const NUM_FIXED: usize = 6;
+        let (fixed_ais, remaining_ais) = array_refs![accounts, NUM_FIXED; ..;];
+        let [
+            mango_group_ai,           // read
+            signer_ai,                // read
+            source_token_ai,          // write
+            destination_token_ai,     // write
+            token_prog_ai,            // read
+            flash_loan_prog_ai,       // read
+        ] = fixed_ais;
+
+        check!(program_id != flash_loan_prog_ai.key, MangoErrorCode::InvalidProgramId)?;
+
+        let mango_group = MangoGroup::load_checked(mango_group_ai, program_id)?;
+        let source_token = Account::unpack(&source_token_ai.try_borrow_data()?)?;
+
+        let flash_loan_amount = if liquidity_amount == u64::MAX {
+            source_token.amount
+        } else {
+            min(source_token.amount, liquidity_amount)
+        };
+
+        let source_balance_before = source_token.amount;
+
+        let mut flash_loan_instruction_accounts = vec![
+            AccountMeta::new(*destination_token_ai.key, false),
+            AccountMeta::new(*source_token_ai.key, false),
+            AccountMeta::new_readonly(*token_prog_ai.key, false),
+        ];
+        let mut flash_loan_instruction_account_infos = vec![
+            destination_token_ai.clone(),
+            source_token_ai.clone(),
+            token_prog_ai.clone(),
+            flash_loan_prog_ai.clone(),
+        ];
+        for account_info in remaining_ais.iter() {
+            flash_loan_instruction_accounts.push(AccountMeta {
+                pubkey: *account_info.key,
+                is_signer: account_info.is_signer,
+                is_writable: account_info.is_writable,
+            });
+            flash_loan_instruction_account_infos.push(account_info.clone());
+        }
+
+        let signers_seeds = gen_signer_seeds(&mango_group.signer_nonce, mango_group_ai.key);
+        invoke_transfer(
+            token_prog_ai,
+            source_token_ai,
+            destination_token_ai,
+            signer_ai,
+            &[&signers_seeds],
+            flash_loan_amount,
+        )?;
+
+        solana_program::program::invoke(
+            &Instruction {
+                program_id: *flash_loan_prog_ai.key,
+                accounts: flash_loan_instruction_accounts,
+                data: cpi_data,
```

Comment:
Above it's suggested that the first 8 bytes of argument to the instruction would be the amount that needs to be repaid. I don't see that here. Communicating it to the instruction is important for the liquidity_amount = u64::MAX case -- the callee has no idea about how much needs to be repaid otherwise (unless the dest account had a zero balance before, I guess)

Reply:
Similarly, the doc comment about starting with a zero byte seems no longer correct.